refactor: Made DatabaseStorageProof stateful

18aaddy opened the pull request:
Resolves #15477

18aaddy commented:
@Rjected @mattsse Can you review pls?

Rjected commented on crates/trie/trie/src/proof/mod.rs:
```diff
@@ -188,9 +188,9 @@ where
     /// The factory for traversing trie nodes.
-    trie_cursor_factory: T,
+    pub trie_cursor_factory: T,
     /// The factory for hashed cursors.
-    hashed_cursor_factory: H,
+    pub hashed_cursor_factory: H,
```
do these need to be pub?

Rjected commented on crates/storage/provider/src/providers/state/latest.rs:
```diff
@@ -118,8 +127,17 @@ impl<Provider: DBProvider + StateCommitmentProvider> StorageRootProvider
-        StorageProof::overlay_storage_multiproof(self.tx(), address, slots, hashed_storage)
-            .map_err(ProviderError::from)
+        StorageProof::overlay_storage_multiproof(
+            &StorageProof::new(
+                DatabaseTrieCursorFactory::new(self.tx()),
+                DatabaseHashedCursorFactory::new(self.tx()),
+                address,
+            ),
+            address,
+            slots,
+            hashed_storage,
```
imo we should implement a method on `StorageProof` that is identical to the previous `from_tx` implementation, then we should be able to do something like:
```
StorageProof::from_tx(self.tx()).overlay_storage_multiproof(address, slots, hashed_storage)
```

Rjected reviewed (changes requested):
I have some suggestions that should simplify how we use these methods, lmk if you have any questions or are stuck with anything

18aaddy commented:
@Rjected , I have updated my code with the requested changes. Here is an explainer:
- I have added a new trait `StorageProofFromTx` with a function `from_tx()` that gives a `StorageProof` instance.

- The reason for defining a new trait is that `StorageProof` is used in the `trie/db/src/proof.rs` file and the `DatabaseTrieCursorFactory` and `DatabaseHashedCursorFactory` structs required for implementing `from_tx()` are also present in `trie/db/src`. If I were to implement the `from_tx()` function for `StorageProof`, it would cause a circular import.

- The `trie_cursor_factory` and `hashed_cursor_factory` fields are still public. The reason behind this is that in the `with_hashed_cursor_factory()` function below, the `StorageProof` instance is consumed, therefore we need to make a new instance. The `StorageProof` struct doesn't implement the Copy trait, and therefore a simple `.clone()` doesn't work.


https://github.com/paradigmxyz/reth/blob/97518b3ccdf8f3ae791c5dfde9dd957b32432037/crates/trie/db/src/proof.rs#L119-L125


This is the best solution I was able to find, if there was something more obvious possible, my apologies. Would love some feedback.

18aaddy commented:
@Rjected @mattsse Can you pls review?

jenpaff commented:
@Rjected could you ptal?
